File: demo-service-impl/src/main/java/com/github/zubmike/service/demo/conf/HibernateFactory.java
package com.github.zubmike.service.demo.conf;

import com.github.zubmike.service.demo.types.*;
import org.hibernate.SessionFactory;
import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
import org.hibernate.cfg.Configuration;
import com.github.zubmike.service.conf.DataBaseProperties;

public class HibernateFactory {

	private static final Class<?>[] ENTITIES = new Class[] {
			User.class,
			Zone.class,
			ZoneSpace.class,
			Starship.class,
			PlanetarySystem.class
	};

	public static SessionFactory createSessionFactory(DataBaseProperties dataBaseProperties) {
		Configuration hibernateConfiguration = createConfiguration(dataBaseProperties, ENTITIES);
		StandardServiceRegistryBuilder builder = new StandardServiceRegistryBuilder()
				.applySettings(hibernateConfiguration.getProperties());
		return hibernateConfiguration.buildSessionFactory(builder.build());
	}

	private static Configuration createConfiguration(DataBaseProperties dataBaseProperties, Class<?>... entities) {
		Configuration configuration = new Configuration();
		configuration.setProperty("hibernate.connection.driver_class", dataBaseProperties.getDriverClass());
		configuration.setProperty("hibernate.dialect", dataBaseProperties.getDialect());
		configuration.setProperty("hibernate.connection.url", dataBaseProperties.getUrl());
		if (dataBaseProperties.getUser() != null) {
			configuration.setProperty("hibernate.connection.username", dataBaseProperties.getUser());
		}
		if (dataBaseProperties.getPassword() != null) {
			configuration.setProperty("hibernate.connection.password", dataBaseProperties.getPassword());
		}
		dataBaseProperties.getProperties().forEach(configuration::setProperty);
		for (Class<?> entity : entities) {
			configuration.addAnnotatedClass(entity);
		}
		return configuration;
	}

}

File: demo-service-kotlin-impl/src/main/kotlin/com/github/zubmike/service/demo/conf/HibernateFactory.kt
package com.github.zubmike.service.demo.conf

import org.hibernate.SessionFactory
import org.hibernate.boot.registry.StandardServiceRegistryBuilder
import org.hibernate.cfg.Configuration
import com.github.zubmike.service.demo.types.PlanetarySystem
import com.github.zubmike.service.demo.types.Starship
import com.github.zubmike.service.demo.types.Zone
import com.github.zubmike.service.demo.types.ZoneSpace
import com.github.zubmike.service.conf.DataBaseProperties

object HibernateFactory {

    private val entities: List<Class<*>> = listOf(
            Zone::class.java,
            ZoneSpace::class.java,
            Starship::class.java,
            PlanetarySystem::class.java)

    fun createSessionFactory(dataBaseProperties: DataBaseProperties): SessionFactory {
        val hibernateConfiguration = createConfiguration(dataBaseProperties, entities.toList())
        val builder = StandardServiceRegistryBuilder()
                .applySettings(hibernateConfiguration.properties)
        return hibernateConfiguration.buildSessionFactory(builder.build())
    }

    private fun createConfiguration(dataBaseProperties: DataBaseProperties, entities: Collection<Class<*>>): Configuration {
        val configuration = Configuration()
        configuration.setProperty("hibernate.connection.driver_class", dataBaseProperties.driverClass)
        configuration.setProperty("hibernate.dialect", dataBaseProperties.dialect)
        configuration.setProperty("hibernate.connection.url", dataBaseProperties.url)
        if (dataBaseProperties.user != null) {
            configuration.setProperty("hibernate.connection.username", dataBaseProperties.user)
        }
        if (dataBaseProperties.password != null) {
            configuration.setProperty("hibernate.connection.password", dataBaseProperties.password)
        }
        dataBaseProperties.properties.forEach { (propertyName, value) -> configuration.setProperty(propertyName, value) }
        for (entity in entities) {
            configuration.addAnnotatedClass(entity)
        }
        return configuration
    }

}